{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Maximum Likelihood tree inference"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sequences generated in for this study are in the following files:\n",
    "\n",
    " - `sequences/Gammarus_sp_SANGER.nr.fasta` #Sanger sequences\n",
    " - `../3-extract_haplotypes/Gammarus_sp_from_metaBEAT.fasta` #haplotyes extracted from metabarcoding data\n",
    " \n",
    "Other Gammarid COI sequences from Weiss et al. 2014 are at:\n",
    " - `../1-download_reference/Gammarus_COI_refseqs.gb`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find the Reprophylo manual [here](https://docs.google.com/document/d/1Q-8B0cvkZw2zMkuP0Af4zZ7FiAvBQPDdGbrLLMgtx_4/edit#)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import reprophylo functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from reprophylo import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specifying the locus/loci to be used in the analyses."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Genbank input files may contain sequences for a range of genes (say you have downloaded full mitochondrial genomes), but you are only interested in a particular gene (say COI). \n",
    "\n",
    "So, first we'll want to get an overview of all the genes present in your input Genbank file. Reprophylo has a function for that.\n",
    "\n",
    "The function will write tentative locus descriptions to the file `loci.csv`, which you may modify, e.g. exclude certain genes from the loci file, or synonymize gene names if necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "There are 2 gene names (or gene product names) detected\n",
      "----------------------------------\n",
      "Gene and count sorted by gene name\n",
      "----------------------------------\n",
      "27 instances of CDS,COI\n",
      "72 instances of CDS,cytochrome oxidase subunit 1\n",
      "-------------------------------\n",
      "Gene and count sorted by counts\n",
      "-------------------------------\n",
      "72 instances of CDS,cytochrome oxidase subunit 1\n",
      "27 instances of CDS,COI\n"
     ]
    }
   ],
   "source": [
    "list_loci_in_genbank('../1-download_references/Gammarus_COI_refseqs.gb', 'loci.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function summarizes all gene names detected in the Genbank file. Reprophylo internally already contains a list of synoyms for common genes and attempts to summarize all unique gene names in the file `loci.csv`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dna,CDS,MT-CO1,COI\r\n",
      "dna,CDS,cytochrome_oxidase_subunit_1,cytochrome oxidase subunit 1\r\n"
     ]
    }
   ],
   "source": [
    "!cat loci.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on the file `loci.csv` produced by the function we prepare a file `target_loci.csv` that contains all synonyms for the COI gene that appear in the Genbank file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting target_loci.csv\n"
     ]
    }
   ],
   "source": [
    "%%file target_loci.csv\n",
    "dna,CDS,MT-CO1,COI,1\n",
    "dna,CDS,cytochrome_oxidase_subunit_1,cytochrome oxidase subunit 1,1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Start a new Reprophylo project, specifying/characterizing the loci to be processed via the `loci.csv` file that we have prepared."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:Cloud:Log file (/root/.picloud/cloud.log) opened\n"
     ]
    }
   ],
   "source": [
    "pj = Project('target_loci.csv', pickle='my_project.pickle', git=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check if the Reprophylo project was created correctly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Project object with the loci MT-CO1,\n"
     ]
    }
   ],
   "source": [
    "print pj"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reading in the sequences from Genbank format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/reprophylo/reprophylo.py:1015: UserWarning: Version control off\n",
      "  warnings.warn('Version control off')\n"
     ]
    }
   ],
   "source": [
    "input_files=['../1-download_references/Gammarus_COI_refseqs.gb']\n",
    "pj.read_embl_genbank(input_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All sequences are now stored in memory as 'sequence records'. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many sequences does our dataset contain?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "99\n"
     ]
    }
   ],
   "source": [
    "print len(pj.records)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reading in the new sequences produced from Sanger sequencing or metabarcoding. They will be added to the sequence records with the id 'denovo'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "denovo_sequence_filenames=['sequences/Gammarus_sp_SANGER.nr.fasta', \n",
    " '../3-extract_haplotypes/Gammarus_sp_from_metaBEAT.fasta']\n",
    "\n",
    "pj.read_denovo(denovo_sequence_filenames, 'dna', format='fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "118\n"
     ]
    }
   ],
   "source": [
    "print len(pj.records)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We are assuming that the sequences that came from Genbank are already in the correct orientation, but we should check that our custom sequences need to be inverted. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll use the `blast` program and a few functions to do that."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`blastn` is installed in the image. Otherwise on ubuntu install like so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading package lists... Done\n",
      "Building dependency tree       \n",
      "Reading state information... Done\n",
      "ncbi-blast+ is already the newest version.\n",
      "0 upgraded, 0 newly installed, 0 to remove and 287 not upgraded.\n"
     ]
    }
   ],
   "source": [
    "!apt-get install ncbi-blast+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "!mkdir temp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write all sequences to file `temp.fasta`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "output_handle = open(\"temp/test.fasta\", \"w\")\n",
    "SeqIO.write(pj.records, output_handle, \"fasta\")\n",
    "output_handle.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert `temp.fasta` to blast database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      "\r\n",
      "Building a new DB, current time: 01/13/2017 09:32:29\r\n",
      "New DB name:   temp/all\r\n",
      "New DB title:  temp/test.fasta\r\n",
      "Sequence type: Nucleotide\r\n",
      "Keep Linkouts: T\r\n",
      "Keep MBits: T\r\n",
      "Maximum file size: 1000000000B\r\n",
      "Adding sequences from FASTA; added 118 sequences in 0.00350595 seconds.\r\n"
     ]
    }
   ],
   "source": [
    "!makeblastdb -in temp/test.fasta -dbtype nucl -out temp/all"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perform blast search."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "!blastn -db temp/all -query temp/test.fasta -outfmt '6 qseqid sseqid qstart qend sstart send' -out temp/test.blastn.out"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Identify any denovo sequences that match with non-denovo sequences in reverse complement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Records to invert:\n",
      "['denovo0', 'denovo1', 'denovo2', 'denovo3', 'denovo4']\n"
     ]
    }
   ],
   "source": [
    "blastout=open('temp/test.blastn.out')\n",
    "\n",
    "to_invert=[]\n",
    "for line in blastout:\n",
    "    col=line.strip().split(\"\\t\")\n",
    "    if 'denovo' in col[0]:\n",
    "#        print col\n",
    "        if (int(col[-2]) > int(col[-1])) and (not 'denovo' in col[1]):\n",
    "#            print \"## invert ##\"\n",
    "            to_invert.append(col[0])\n",
    "        \n",
    "to_invert=list(set(to_invert))\n",
    "print \"Records to invert:\\n%s\" %to_invert\n",
    "\n",
    "blastout.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reverse complement records."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are records to invert\n",
      "\n",
      "\t-> denovo0\n",
      "\t-> denovo1\n",
      "\t-> denovo2\n",
      "\t-> denovo3\n",
      "\t-> denovo4\n"
     ]
    }
   ],
   "source": [
    "if to_invert:\n",
    "    print \"There are records to invert\\n\"\n",
    "    for r in pj.records:\n",
    "        if r.id in to_invert:\n",
    "            print \"\\t-> %s\" %r.id\n",
    "            r.seq = r.seq.reverse_complement()\n",
    "            to_invert.remove(r.id)\n",
    "else:\n",
    "    print \"Nothing to invert\\n\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove temporal directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!rm -rf temp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, lets print a summary of the number of sequences per species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.species_vs_loci('species.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "species\tMT-CO1\r",
      "\r\n",
      "Gammarus balcanicus\t1\r",
      "\r\n",
      "Gammarus cf. fossarum A DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum B DCC-2015\t2\r",
      "\r\n",
      "Gammarus cf. fossarum C DCC-2015\t1\r",
      "\r\n",
      "Gammarus cf. fossarum D DCC-2015\t5\r",
      "\r\n",
      "Gammarus cf. fossarum E DCC-2015\t13\r",
      "\r\n",
      "Gammarus cf. fossarum F DCC-2015\t5\r",
      "\r\n",
      "Gammarus cf. fossarum G DCC-2015\t7\r",
      "\r\n",
      "Gammarus cf. fossarum H DCC-2015\t11\r",
      "\r\n",
      "Gammarus cf. fossarum I DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum J DCC-2015\t1\r",
      "\r\n",
      "Gammarus cf. fossarum K DCC-2015\t4\r",
      "\r\n",
      "Gammarus cf. fossarum L DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum M DCC-2015\t5\r",
      "\r\n",
      "Gammarus cf. fossarum N DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum O DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum P DCC-2015\t3\r",
      "\r\n",
      "Gammarus fossarum\t17\r",
      "\r\n",
      "Gammarus glabratus\t1\r",
      "\r\n",
      "Gammarus pulex\t6\r",
      "\r\n",
      "Gammarus roeselii\t1\r",
      "\r\n",
      "Gammarus tigrinus\t1\r",
      "\r\n",
      "undef\t0\r",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!cat species.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some sequences seem to be missing. This is because the `species_by_loci` function only considers sequence records that are identified with the correct gene identifier. We need to add this for our custom sequences so that reprophylo understands that these sequences are to be analysed together with the COI sequences that we have loaded from the Genbank data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for r in pj.records:\n",
    "    if 'denovo' in r.id:\n",
    "        pj.add_feature_to_record(r.id, 'CDS', qualifiers={'gene': 'MT-CO1'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.species_vs_loci('species.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "species\tMT-CO1\r",
      "\r\n",
      "Gammarus balcanicus\t1\r",
      "\r\n",
      "Gammarus cf. fossarum A DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum B DCC-2015\t2\r",
      "\r\n",
      "Gammarus cf. fossarum C DCC-2015\t1\r",
      "\r\n",
      "Gammarus cf. fossarum D DCC-2015\t5\r",
      "\r\n",
      "Gammarus cf. fossarum E DCC-2015\t13\r",
      "\r\n",
      "Gammarus cf. fossarum F DCC-2015\t5\r",
      "\r\n",
      "Gammarus cf. fossarum G DCC-2015\t7\r",
      "\r\n",
      "Gammarus cf. fossarum H DCC-2015\t11\r",
      "\r\n",
      "Gammarus cf. fossarum I DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum J DCC-2015\t1\r",
      "\r\n",
      "Gammarus cf. fossarum K DCC-2015\t4\r",
      "\r\n",
      "Gammarus cf. fossarum L DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum M DCC-2015\t5\r",
      "\r\n",
      "Gammarus cf. fossarum N DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum O DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum P DCC-2015\t3\r",
      "\r\n",
      "Gammarus fossarum\t17\r",
      "\r\n",
      "Gammarus glabratus\t1\r",
      "\r\n",
      "Gammarus pulex\t6\r",
      "\r\n",
      "Gammarus roeselii\t1\r",
      "\r\n",
      "Gammarus tigrinus\t1\r",
      "\r\n",
      "undef\t19\r",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!cat species.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Better, but our custom sequences are now assigned to an undefied species. We'll need to add the species to the record. We will look in the original header of the sequence for the species name and add the appropriate organism name to the sequence record."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for r in pj.records:\n",
    "    source=r.features[0]\n",
    "    if 'denovo' in r.id:\n",
    "        if 'pulex' in source.qualifiers['original_id'][0]:\n",
    "            r.features[0].qualifiers['organism'] = ['Gammarus pulex']\n",
    "        elif 'fossarum' in source.qualifiers['original_id'][0]:\n",
    "            r.features[0].qualifiers['organism'] = ['Gammarus fossarum']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.species_vs_loci('species.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "species\tMT-CO1\r",
      "\r\n",
      "Gammarus balcanicus\t1\r",
      "\r\n",
      "Gammarus cf. fossarum A DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum B DCC-2015\t2\r",
      "\r\n",
      "Gammarus cf. fossarum C DCC-2015\t1\r",
      "\r\n",
      "Gammarus cf. fossarum D DCC-2015\t5\r",
      "\r\n",
      "Gammarus cf. fossarum E DCC-2015\t13\r",
      "\r\n",
      "Gammarus cf. fossarum F DCC-2015\t5\r",
      "\r\n",
      "Gammarus cf. fossarum G DCC-2015\t7\r",
      "\r\n",
      "Gammarus cf. fossarum H DCC-2015\t11\r",
      "\r\n",
      "Gammarus cf. fossarum I DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum J DCC-2015\t1\r",
      "\r\n",
      "Gammarus cf. fossarum K DCC-2015\t4\r",
      "\r\n",
      "Gammarus cf. fossarum L DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum M DCC-2015\t5\r",
      "\r\n",
      "Gammarus cf. fossarum N DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum O DCC-2015\t3\r",
      "\r\n",
      "Gammarus cf. fossarum P DCC-2015\t3\r",
      "\r\n",
      "Gammarus fossarum\t26\r",
      "\r\n",
      "Gammarus glabratus\t1\r",
      "\r\n",
      "Gammarus pulex\t16\r",
      "\r\n",
      "Gammarus roeselii\t1\r",
      "\r\n",
      "Gammarus tigrinus\t1\r",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!cat species.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_Gammarus fossarum_ is considered a complex of cryptic species. To make our lives easier for the analyses we'll correct the tentative species names _Gammarus cf. fossarum_ as used by Copilas-Ciocianu and Petrusek 2015 to simply _G. fossarum_."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for r in pj.records:\n",
    "    source=r.features[0]\n",
    "    if 'Gammarus cf. fossarum' in source.qualifiers['organism'][0]:\n",
    "        #keep the original tentative identification in separate qualifier\n",
    "        source.qualifiers['tentative'] = [source.qualifiers['organism'][0]]\n",
    "        #rename organism to simply Gammarus fossarum\n",
    "        source.qualifiers['organism'] = ['Gammarus fossarum']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.species_vs_loci('species.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "species\tMT-CO1\r",
      "\r\n",
      "Gammarus balcanicus\t1\r",
      "\r\n",
      "Gammarus fossarum\t98\r",
      "\r\n",
      "Gammarus glabratus\t1\r",
      "\r\n",
      "Gammarus pulex\t16\r",
      "\r\n",
      "Gammarus roeselii\t1\r",
      "\r\n",
      "Gammarus tigrinus\t1\r",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!cat species.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There! Now, let's produce some stats for the sequences in our dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.extract_by_locus()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Distribution Of Sequence Lengths\n",
      "Distribution Of Sequence Statistic \"Gc_Content\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/pymodules/python2.7/matplotlib/figure.py:371: UserWarning: matplotlib is currently using a non-GUI backend, so cannot show the figure\n",
      "  \"matplotlib is currently using a non-GUI backend, \"\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAEwAAAI7CAYAAABGL317AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAGapJREFUeJztnX+YFWW9wD9ngd2FBZRNZFEoJN1kVcQyRK2rJaiUP/L2\n89JTUF4v91pWdk3IW9d97Ifpxcrs6rW0kJvSQ3FLLLKg8sfTY2mGuAmroRCi/FhFCRB2Wc65f3zf\n05k9OzO738M5Z+bMfD/PM8+Z886c4d0P78y88877fl8wDMMwDMNIBG8C1niWXcCngGZgFfAM8Cvg\ncM9vPg/8BegEzq1mZuNGHbAVmAjcCFzt0hcAX3PrbcATwDBgErDB/S6VnAs87NY7gXFuvcV9Byld\nCzy/uR+YUZXcOeL0v/MhYKlbHwdsd+vbKcg7Ctji+c0W4Oiq5M4RF2H1wIXAj3y25dwSRNi2shMX\nYbOBx4Eu9307cioCjAd2uPUXkGtcngku7e80NTVtoCC51OWJsv1lFeKHwFzP9xspXKsW0v+iXw8c\nAzwLZIqONZgS1z7A9qqWWi1NwEvAKE9aM7Aa/2rFNcjdsRM4z+d4iRdWbioqLC7XsGrzQNQZiBPl\nOJ2shJULE6bEhCkxYUpMmBITpsSEKTFhSkyYEhOmxIQpMWFKTJgSE6bEhCkxYUpMmBITpsSEKTFh\nSkyYEhOmxIQpMWFKTJgSE6bEhCkxYUpMmBITpsSEKTFhSkyYEhOmxIQpMWFKTJgSE6bEhCkZGnUG\nqkl7e3vOs148RmlQWAlTkiphpZYqL6kSVg5MmJI4CDsc+DGwHliHDHpvR8Zz58MzzPbsH2kYhjjc\nJW8GVgLvQ/LThAwc/bpbvLQBH3SfRyODUFuBbLUyG3UJOwx4O/A9970XCfYB/YcmA1yMRB44AGxC\nRuZOr2wW+xK1sGOQgfHfB/4EfBcY4bZdAawF7qQwhDn1YRiGAm8GbnWfe5FB8bciMqch0VJuCjlG\nVcdnR30N2+KWx9z3HyPCujz73AHc59YHDMMAMH369OWe9XXz589fn/++fLls6ujomJNPu/3226c8\n+uijbfnvjz76aGl/TZV4CLlwg9wdb6AQswLgSuAet37IYRja29tz3kekAAK3R13CQK5VdyMSngU+\nDnwLOR1zwEZgvtt3HbDMffYCl5OyUxLkwv7WorSPhuz/VbdEQtQX/ZrDhCkxYUpMmBITpsSEKTFh\nSkyYEhOmxIQpMWFKTJgSE6bEhCkxYUpMmBITpsSEKTFhSkyYEhOmxIQpMWFKTJgSE6bEhCkxYUpM\nmBITpsSEKTFhSkyYEhOmxIQpMWFKTJgSE6bEhCkxYUpSJWwUPOW3ruGQx0DHkBwBf1czbH8ZXgV4\nHRy+szBZ8aCPkaoSVg5MmBITpsSEKTFhSkyYkjgIK45bcRoyz/cq/Of5jjRuRRy4CxmFCzLg9TBk\nJvmrXdoC+s8kPwyYhIRhKP5PDxyh2wzbc/B0Dp5uhu0heYrtxOiHAc/5pHdSqFS2uO8gpWuBZ7/7\nkUgqXioqLOpT0i9uRRMiK/8HbacgL/K4FVGP+c7HrfgkEorhm0gYBi85wk+RftuCwjDMestbGg/W\n148CqO/paex4/PE5UFthGFqQqAF53gb8HLkB5EMxjKdwSi6kr9D7kZuEF1+5OR9C8hXbU3Ib8DyF\nuBUzkVaE+4C5Lm0u8FO3vgL4EIW4FccB8S0OFeJk5HRcC/wfciNoRiI3+VUrrkHujp1IFKhiKnrR\nt+Yd5TGiPiVrDhOmxIQpMWFKTJgSE6bEhCkxYUpMmBITpsSEKTFhSkyYEhOmxIQpMWFKTJgSE6bE\nhCkxYUpMmBITpsSEKTFhSkyYEhOmxIQpMWFKTJgSE6bEhCkxYUpMmBITpsSEKTFhSkyYEhOmxIQp\nMWFKTJgSE6bEhCkxYUpMmBITpiQOwjYBTwJrKAwWbUfGc69xy2zP/pGGYYh6zDfI2MSzgZ1FaV93\ni5c24IPu82hkEGorkK14Lh1xKGHgP5jTL+1iYClwACmZG4DplctWf+IgLIeUlD8Cl3nSr0CGNd9J\nYQhz6sMwAJwJbAXGIuFjOoHbgOvc9i8BNwGXBvw+VWEYirkW+PeitElAh1svOQwDJCMyyghglFtv\nQu56HRRiVgBcQkFY5GEYoj4lxwE/cetDgbuRsAtLgGnI//RGYL7bZx2wzH32ApdT5cBCFoZBeYyo\nT8maw4QpMWFKTJgSE6bEhCkxYUpMmBITpsSEKSnlWXIMcCTy+NAFvFLWHMWcwQgbgrR0vh84i74t\nCSBtWQ8CPwLupYrNxVEQJmwo8G9I+9N4YC/SKroSeBl5OG0GjgUuQppdtgLXIw2AByuW65jyNCLp\nTuQlxZCQfYcA70Qi/u6lEHc1CiKLRf0Ngps/wmhBIv5GRaiw9vb2XHt7e65UYWGn5JWDyZ0P24DP\nlPjbinIAXkJuWPl1NamqVuyGE/zWNZRSrTgNaWc/xn1/Dgl9/IdSMlBraIQNQcK3z/PZtgBph7+U\nhN8dNafkFxBZPwXOQCqwY5D3ivcCHwW+WOb8xQ6NsI8jb6j/Efg9sMstj7i0XwMfK3cG44ZG2JFI\nSfIj57aVUg2pKTTC/kL/xyIvLUhlN9FohF2PTEUxzWfbKcAn3D6JRnOXbEWqEI8hnUbWu/Q2ZKaF\nJ90+/1n0u+tIEBph13rWz3eLl1PcUkxqhU2uWC5qCI2wTZXKRC1Rau+dEcAb3PpfgdfKk534o334\nPgH4BVJhfcotr7q0E8ubtXiiKWGnIE3RTfS/S56LPCKdhXQTTywaYf+FtNe/FZlHzcubgd+6fWaW\nJ2vxRHNKzgC+TX9ZuLRvA6eXI1NxRiNsP/KSI4itwL5Dy0780QhbibwdCuJCt0+i0Qj7LPA6ZHrW\n6Ujv51FIC+xy5JVbqe8Baoawi36Wvp1j8+tvRtq//NhB+Ou4midM2JISjhfbuWXLRZiwedXKRC2R\nqtds5SBM2JhDOO6h/DbWhAnbhLSBvU5xvCOR0WcbB9qxVgkTthBpdn4RGQ90GTI960jPPqOQu+Yn\nkMmAX3D7LSChhF30bwPuQQZAzUf6iOU54D6HedKeQ+av/R9gdxnzGCsGevjehbzYuAGprJ6FtE6M\npdAD8c/Ig/fjlctmfBhsa0UWeXn7+wrmpSaIQ7ViE/3DMDQjbW5+01ZHGoYhDsLyYRhOoRAhYCEi\nrBXpgpAftuwNw3A+cCtV/hviIAz6D+a8CLjLrd8FvMetWxgG/MMwjKPQpXI7hT4bFoYB/zAMXnKE\nP9QPOgwDwPLlsqmjo2NOPi0JYRg6KXR8GU9B4iGFYQDIdwoeIB81F4ZhBTDXpc9FOvFBDYZhqEPe\nCh2LPGP6jbzX9KUICsPwRyTcwqXIxf0Dbp/IwzBohB2HdJo7foD9NMI24t99aifBr+u+6pZI0Ai7\nBemQcjXyKPRyRXIUczTC3g7cDCyqUF5qAs1FvxtpkUg1GmG/ROpMqUb7XvJ04Crktp5Kwq5hG+l/\nyx4J3Ii0kb1I31EfGbd/onsqhgn7awnHS/V7ybOrlYlaQnMN+wfkATmIsW6fRKMR9gAwK2T7OUiF\nNtGU8+F7CCm4hpVT2OmUOCy4lhjo0ejTyPjtfMn5JvBln/2agdHA98qXtXgymPeS+erFJKQE7Sja\nJ4d0P38EiUSQaAYSttgtIO1Snyd4zGQq0LRWTKpUJmqJqJuoaw6NsCzy7Bi27EGaj29GXoklDo2w\nJcgLigwyVHmFW552aR3IW5yDSGj3J0jgg7hW2CTg3cir+kvc0gZcgLzFuQ04CemzfzjSuS5RaIR9\nBQn08QufbSuB77h9QDrXfQ+J+JQoNMJOJryJeiMw1fP9CXTdPWsCjbBdhI9UOwf4m+f7aPebRKER\nthS5Zt0OvAl52K5D3lPe7rYt9ez/DuSOmSg0FdcvIKIuc0u+eTo/VOaXwH+49UbklHyoDHmMFRph\nryF3yHchd8F8OKxNSPXCO5JtP/IYlThK6e60khQM8wvCHo2UaEvYG5A++2G9dxJX9/KiETYb6ac1\nDHlm3OmzT+KbqDXCrkcaEC9G+m+lEs017HikiTq1skAn7CWkB0+q0bZWvLdSGakVNNewxcjjzgqk\ngfA5/EOQbj70bMUXjTBv//kLAvbJkeKoAsUMprOvVSs8tFcqE7WEPRop0QobjQxv+R0yZjEfzekI\nJLrmQH34ax7NKTkWEXUM8CzwRmC42/YyMsRlDAmPv6MR9mVkqMsMpL+Ft49FDqluJPrBG3Sn5AXI\na7SgwfDPAa8/5BzFHI2wI5DrVhBZpGk60WiEbUeuW0FMI+G1fNAJ+zkyHM+vz8RpyAQEie8KpRF2\nHTJG8U8Uht/NBX4IPIwMQ76hhDwMQUIw3Oe+tyNjude4ZbZn30hDMIDuLrkVqXfdgpQ0gI8gd8iV\nyCxbpQwJ/DTy/jI/MjcHfN0tXrwhGI5GBta3UuWpzbQV181Ii2szUr04HYnodCF9R/sPlgnIa7s7\nKLwfyOD/riDyEAxQ+qPRLmSs9R84tIGm3wA+R99SkkO6S61FpjnLR0WJPAQDRPsseQFS+V1D3xJ1\nG/I0MQ25DNwUcoyqt45oomyGkd9P0x52BhIB5V1I/W000qr7Uc8+d1C4GbwATPRsm+DS+lHJuBVh\nMhaHbAsiR2lT9ZyFjMO8EIlTkY9IfCUS+3oOcrG/B7lu5S/6x9K/lIX+J+djVrS3t4f97YHHiEuU\nzXzpBBmPebL7vhF5cQwxCMEA8QglAzLw6wG3/pGQ/SINwQDWgKjGhCkxYUpMmBITpsSEKdFUK0od\nAJ+ojsEaYQ+4T79acFDNOHFdBzTCLkVaEY5DAqN55zWag8RcvaWsuYshGmGjkXeTx9P/ofdLyBDm\n0cC3ypO1eKK56H8SGfHh10KwxW37ZDkyFWc0wiYCe0O2v4a9l+zDJuTB2O/d43C3bdOhZyneaK5h\nNyGn3R+B/6bQwW4KMgHBFOBfy5q7GKIR9l0k1upXEGFe9iENgN8pU75ii7Y97JtIS+y5FMZzP4uE\nRH61fNmKL6U0IL6KtHymklKEHYOMzD0SqcBuQmIitiD9LxLdl1/78H0j8qr+dqTrQP60HI7U/C8v\nX9biiUbYfOTC/m3kGuZ9dtyFdEQJ6o6eGDSn5OXIaLbPIH3FiulAXpfFlkGEhR8QTQlrRSKRB9GF\nv8jU8goyOQCImCx9+7ReR2Faiigpx7vKskxA8BgSasGPRuTR6HeK4yWemUip+gFSsrLAh5Hpcv6A\nvI0+I7LcFahoCdPyL0iIhWzRso/4TOBZUWGD6ZlTzHjgfcjDdgZpaV1GQE+aCBhsj6NKH6NmqGgJ\nO5TOKMOQrkhHIz1qnjqEYyWGs5E2+nFF6ccgE9Dlr2EHge9XNWfBRHrRX4xco4p5EBH1MNKw2OG+\nzyv1HyojkQpbhzw7ejkekfOgJ204Egvx16X+Q2Uk0oprC/1L2Nnu8w5P2j6kO+VUEs5AwhoQGV7y\nfeMfLEp/nr4TZyaSgYQ9D5xQlPY2pLt48UCsEaSgmXogYQ8h3cBPct8vQXou+0XaPJH4VF4jYzLy\n8vYgUqqySBP0iUX7DUFmoSm+QURB5M+SpyKDCzqBnyFjjIqZhUxfHTYFRrWIXFitEZv2MAMTpsaE\nKTFhSkyYEhOmxIQpiYOw4jAMzUj3qWeQF8feB/rIwzDEQVg+DEO+srgQEdaKtK/lZ473hmE4H7iV\neOS/qkxAhiK/g0IJ66TQJN5CoWvo54EFnt/ej/9jWqJr+n5hGMZR6HKwnYI8C8OAfxgGLznCS0yq\nxnz7hWH4X6RUtQDbkJfG+cBuZQnD4IcmDENcOIvCNexGCteqhcDX3HobMm1GPYWwgkEDwg6V2Dfv\nnIWEBASpVqzGv1pxDRJzpxM4L+BYqRBWThJ9l6w5TJgSE6bEhCkxYUpMmBITpsSEKTFhSuIScK0q\nTG2c0NrW0DKzqa6hfm+2u2dd97bVT+7f4tfDMpDUCJvaOKF11sgp864aO6snn7aoa9U8YLFGWmpO\nybaGlpleWQBXjZ3VM6WhJWwa236kRlhTXUO9X/rIgPQgUiNsb7a7xy99T0B6EKkRtq572+pFXav6\nlKZFXavq13dvW605ThLH0wSOE5raOKF1SkPLzJF1DfV7st0964PvkjbWqFzHSM0pWS5MmBITpsSE\nKTFhSkyYEhOmxIQpMWFKTJgSE6bEhCkxYUpMmBITpsSEKTFhSkyYEhOmxIQpMWFKTJgSE6bEhCkx\nYUpMmBITpsSEKTFhSkyYkiiFNSJR0p9AwjBc79LbkQHwa9wy2/ObyONWRM0I9zkU+D0SkPJa4LM+\n++aHMA8DJiEjc/3+wxPdP+w191mPREh5xX336/13MbAUOIBMDbSBQojUqhG1sDqk1GwHfkshAPgV\nwFrgTgpjvmMRtyLqgQ1ZYBpwGPBLJArxbcj8IiATfd6EzJ7qh++pk4YwDABfROZN8jIJCQwOEpJh\noWfb/cBpPsdJbFSBIyicbsORILvnIEE+8lyJxLiGmMStiPKUHA/chVzH6pCoKL8GliCnaQ7YiMzY\nBVL1WOY+e5GJqWJZEmqNRFcrag4TpsSEKTFhSkyYEhOmxIQpMWFKTJgSE6bEhCkxYUpMmBITpsSE\nKTFhSkyYEhOmxIQpMWFKTJgSE6bEhCkxYUpMmBITpsSEKTFhSkyYEhOmxIQpMWFKTJiSqHtRV5UT\nhx21rG34+Iua6urr9mZ7suv2bV3x5wMvfkBzjNQIO3HYUcvObJr83pmj2zIZMuTIsbpu3XvZyzKN\ntNSckm9sGHvRtBETM0AmI72vM9NGTMxMbjjiIs1xUlPC6jN1Q8cMacqcMWLy39Me3rshU58ZonKQ\nmhI2JFOXObPp2L9PiZoDzmw6lqGZOlVU89QI25/r3fvigVf7pL144FX253r3ao6TmlOyN9fbs7H7\npRyQqSNDlhwbu1/K9eZ6uzXHSY2wYZmhQ3fn9h/MkavLAjly7M7tzw7LDB2mOU5qhI2qG76vrfGo\nxs0HdubqyGSy5HJtjUdlH97zl32a46RG2O5cd2djZuipM0ZMHoJUK3JdvbsP7s7u7xzot15Sc9Hf\n0L3j+sU7H9m8o3f37q7e3Xt29O7evXjnI5s39Lx0/cC/LpDEiUICJ0A5qeHoc6c0tnxsZF1D455s\n9/71+7d9v6P7hV9pjpFEEjuaLSgMQzOwCv95vi0Mg/v0hmG4EbjapS+g/0zykYZhiAsjgMeAE5DS\nM86lt7jvIKVrgec39wMzfI6V2FMy/+8Xh2EY577jPvPyLAwD/cMwvKNoe/45OQgLw4CcgvnIAuMp\nnJKxCMMQ5SlZHIZhFhKcaAUw16XPBX7q1lcAH6IQhuE4oDaKQpk4CfgTcg17EvicS28GVuNfrbgG\nuTt2AucFHDcVd8lykthTsiYxYUpMmBITpsSEKTFhSkyYEhOmxIQpMWFKTJgSE6bEhCkxYUpMmBIT\npsSEKTFhSkyYEhOmxIQpMWFKou4qUFWmNk5obWtomdlU11C/N9vds6572+on9295RnOM1Aib2jih\nddbIKfOuGjurJ5+2qGvVPGCxRlpqTsm2hpaZXlkAV42d1TOloWWm5jipEdZU11Dvlz4yID2I1Ajb\nm+3u8UvfE5AeRGqErevetnpR16o+pWlR16r69d3bVmuOk8Su1YFdxqc2Tmid0tAyc2RdQ/2ebHfP\n+uC7pHU7L9cxUnNKlgsTpsSEKTFhSkyYEhOmxIQpMWFKTJgSE6bEhCkxYUpMmBITpsSEKTFhSqIW\nNpHCWO8/A59y6e3ImO41bpnt+U05QjGcXeLvIqcFGfMNMBJ4GpgCXAt81mf/wYRiGEyLa/sA22Pb\n4roNEQCwB1hPIVKAX5v6xcBS4ACwCRE2vbJZ7EvUwrxMAk5BAn4AXAGsBe6kMIw58lAMcRE2Evgx\n8GmkpN2GDISfBmwFbgr5bfHpsxb6hDr0W64dYPvaoH8sDn0rhgHLgR9QiCCww7P9DuA+t/4CcqPI\nM8GleZlGgskAS4BvFKWP96xfCdzj1vMX/XwohmdJ0ftDkOBEWUSCtwqxBAnNsBYpdeM8vxlMKAbD\nMGqOY4HfRJ2JWmIacv1UEYdqRaXI17WCGB+yLZAk35KzwEbgtYDtw5GqyZCq5SjmPAv8U8j2kk7J\nuDwaVYI1yLOpMUhOAE4N2Z5vIjIqSZIv+l6mAq1u/Rnkscvw4VSgA7m4e5cngbdEmK9Y8ibgb8Dj\nwBzkBnAK8GEk9uIut4/hWIpE5vS77NS5bUurmqOYsx3/OK95ZtC3oTL1dCMtskFMdPuoSHLFdQtw\ncsj2k+nfvJ1qFiEviFt8to1328JerviS5HrY4cAjyGl5N/LOE+QJYA5SumYAr0SSu5gyBrgV2Emh\nDrbTpY0p5YBJLmFe6oCxbr2LElop0sAoJFr6YT7bDnPbRlY1RzHnc8DKkO0/Q1plDcfjwLtCtp+P\nPCIZjl3A60O2v8HtoyLJFdcMcGTI9iMp4e9PsrCnCO9KcJ7bx3D8M/LG6D0+2y5x2y6rao5iTgbp\nQpVFavk/ccvTLs2adgL4AHAvMv/beqQ3kGoydMMYND+nxC4CaWU3MHnAvUJIcrWiIiRZ2G/oO+sW\nwGakj7/hQ5bwmn5JJLmEVYQkd6gDeTPUOMA+mzUHTHKL62BaVXMoO9QlvYSdj7ThG4PALvpxIMnC\n/gocLPdBk3zRv4/g4I/59BxwkeagSb7ovxupMjxAuDjDcQMyOPU54AuE9+QxHEORceL3AvuBXwDv\nR3pQGwPQAixAxll2UeJb7yTfJYtpQroIjELaxQwfRgDzgIeQt0R3A+dEmaE4cweFXtSX079trCSS\nXA/LAs9TGMTgrUJkPGlWD3MswV+SF6uHGYZhGEbZ+H8g+FdobqlGRQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7feb58f22190>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAEUAAAI7CAYAAAC6OlaxAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFCxJREFUeJztnXu0HVV9xz83PMyDAHmSWN6YoNgFWGhAWgQiCIQWoW2w\nhhaQqhRZQhEDlsfyWKs8hKqlQmkRBXwsKBEqRSTyiEgLpgVSLpBH8+BNwiMhpA0kkHv7x2+Od+65\n33PmzDn3zJ6Z8/usNSvJzJy5v/vJzD579t6/vcFxHMdxHAeArYDHgTtj+z4PLAaeBC7POqCts/6B\ngnOAp4Gx0b+PAI4H9gXeASYFiisYOwP3YiKqd8qtwMxgEQEjQv5w4JvAXKAvtm8a8BHgEWABcGDW\nQYWU8gfAK1h50hPbvzUwDjgYE3Zr9qGF4+vA88Aq4GXg/4CbgbuBw2LnLQcm1H54zJgxy4H+NrdF\nw/9rDR+HMVCmnAF8Jfr7dOC5Op/pT7hmpYmfK6+Rh2+fKtUAb4i2XmAzcEqwiApIx+6U0N8+nWRB\n6ABCkHSntHyNMt8pLeNSBC5F4FIELkXgUgQuReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkXgUgQu\nReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkXgUgQuRZCn4V3DRqVS6Y/9vafRuQq/UwSlltLKXQIl\nl9IqLkXgUgQuReBSBC5F4FIELkXgUgQuRZAHKbUpuBXghWjf48AxWQeUh7fk2hTcfuDvoi0Ioe+U\nnYFZwPUMJFf2MDjRMnNCS1EpuP1YBvt/A98Fdsw6qDym4F4L7AHsj2WhXpV1YCHLlEOw9P1ZwEhg\ne+AmBidSXs/gORAGMWPGjHmxvz99xhlnLAaYN8929/b2zomff911131g4cKF+1T/vXDhwrZ/iU4S\nT8GdGtt/LvCjOp+pmy5XqVT6402SDch1Cm4PAwFeAewX/XsVlqecKXmRsoCB7NA/DxeGEfrbJ5e4\nFIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSB\nSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsR\n5EGKWhoY4Dws4XJ81gHlQUo1LzmeurYLcBTwbIiAQktReclgidrnB4mI8FJUXvLHsbT+J4JERNgc\nwnhe8uHRvtHAhdijU6VuNnsZU3DV0sC3AWuifauw5cafASaLz5cyBffCaAPLS/4i8Cc156wCDgDW\nZhhX8DIljvpfG471S1OTl7zkX0ZbLXtmHQjk607JDS5F4FIELkXgUgQuReBSBC5F4FIELkXgUgQu\nReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkXg\nUgQuRVBqKeNhzVh4Ku3ngq7N1Sb91Il/PKx5Hd4AmAA7roWd0lyj1HdKq7gUgUsRuBSBSxG4FIFL\nEbgUQR6k1KbgfhVb13QRcB+WeZopeZBSm4JbXZxxf+AO4MtZBxRaikrB3RA7vh3wWtZBhc4Mq6bg\nbl+z/2vYypUbgYOzDipvKbhVLoq2L2HiPqUuUC8F96gDDhi5ZdttxwJsu3nzyN5HH50DxU3Bvanm\nnF2BJ+t8vm7S5XhY0w9L+2HpeEvprUeQxM1miS8NPC22//PAzXU+0zEpocuUKvGlgS8F9ga2ACuA\nM0MEU1S8kSlLXIrApQhcisClCFyKwKUIXIrApQhcisClCFyKwKUIXIrApQhciiBty9torFVsMtZA\n8yqwFGt1Lw3NSBkPnAbMxqYsrP3MO8CjwL8A3wfWDV94YWjUHLkjcAnwOeA92B3xCNZu+nr02fHA\n+7C+menA28A1WNfn+o5FbXSsObIRrwEvYmL2aOL8vbAuzpfIplcvSBfHOdgdkpaRwF+18Lm0dEyK\nt+Z7a35ztCNlNvBzbETzfKxDvBS0KuUs4EasnvJTYDP2dXzB8ISVb+pJexL405p938Tmo86KjhW0\nSXfKY1iFrZYx2DTucV7EBtkUniQpDwL/AVwFjIrtvxN7XE7HpnM/D7iYoUtHlJaDsIn9VwJHR/tG\nAddi7zx9WE32e8DYDOMKPj5lG2xk0UbgB8DEaP8I7OVwq3Yu3iLBpVSZDizAqvGnDMcF20D+Qv2C\ntNdohR7g09gE//Np7r2oEwSV8kfYI/MTbBWEbaP9U4BbsbFqc8m+dhzs8TkbK0gXALdjZcrPas75\nQ2xA32PA77TyQ1okmJQVwD/E/n0cJql25YOxwNXAplZ+SIsEq7xNZPCQzWrmZu2SVRuwkYyHJVyv\nECQ1Rz4MnIuNnV+HNTi9CSypc/4jwxdaftkLq7j1Rdtr2EpNeSDYONoVWDbFNKwWuwSrvZaaZlrz\n+7BG666hUUE7ro3rtvPZ4DSS8gzWOj8hxfUmY90bq9qIKdeciaWebMIqbp/BMrbibSZjsQrbWcBd\nWMfYauCMDOILVnnbAfhr7K7pi22boi2+bzlW3c+q+SB4F8cIYAZWOdsHmMRAX/KTwANY12mWdKyL\no9kO9j6sYtYVlbO89PvUpuF+A1iMpeL+BHuMMyMvUmrTcOcDH8QK9mVYuZYZeZCi0nB/wcAisL+O\nzsmMPEhRK+HGOZ2hbTgdJXQOYaM0XLDG8s3Aj9SHy5iCC43TcE8D/h0b2qHIRWv+97A+oHrMAG5o\n4/rxNNxjsAatifVPD9fyFudUrH2lHnti/7utEk/DvRp7nfgF9mhd08Z1UzOcZcoY7N2nVRZEGwxO\n2M6cJCm7RVv1q/IDwEfEeROwF8jlwxdafqkw+KWv0fYu2fYaBmuOvIOBMSc3AP/E0PeffuB/gYXY\nN0nhSZKyKNoAdgfmAb2dDCgPpCloK50KIm+k/fYZARyJjbKegG7P+Jt2gyoS07A32aQCNytyMX/K\n1VgF7Xyspe31FJ8tFGmkHAp8G7iyQ7HkhjTV/E3YuLfSk0bKPcDvdSqQPJFGyheADzN4NFMpSZPF\nsQp76ZuITSL1UvRn/Fr9DB3Q0ymCd3EAPNsokNgPKTxppBzeqSDyRh4arnNHK1IOwya8/Gfg/dG+\n7bB2lkIPwWiFrbBxs9Xq/BZgZnRsFDbY+KIM48lFG+0F2EDjL2AtcPEC9y1suMaxKa6XW9JIOQWb\n3PJb6PeeJdjbc+FJI2V3LPenHm9QkjIljZQNDB1UHGcvbLxK4Ukj5SHgz+p8ZhzW5/vAcAQVmjRS\nvobl+9yP9QGDjbH9S6zDajvgsmGNriAchw30q21tWw18LONYctHyBjYCcncsmbL6tbwMa1Yo1Rwq\nzbIrNqlMPUZH52RFLipvzwAnNDh+PCUZVDycL4Slebkczl/k/UQNO0UnqaA9NdqqXIRlmtYyAfht\n7P2n8CRJGcfg5sVJWJNknGoH+3fJ9i05F/QBJ4cOIkYu6imlKUiT6JpfNA1ppXwSaz54lcHV/C2x\nPwtPmsdnLnA5lnH6CLqhqeu6OM7CxsnPxJofS0uax2cK1hxZaiGQTsoKbD7J0pNGypVYbTbLKYaC\nkKZM6cMqQouxcfor0d82tet+FY40ow6aGc/WT3bzM+Vi1MHM5FPKQRopCzoVRN5op5o/kcb5OIUl\nrZTfwgrS9Via2yvYZDM3RsdapTYFdzaWBLWFbOd5AtI9PrtiNdqdsHzh6lRE+2DTrh6FZY61krRQ\nTcGtft33AicC17VwrUy5EavNzhLHjsUmm7mxhevuDNwLHMHQuScfoP6dkovW/I9h80WqdNi7sZS2\no8WxJJJScDMnjZRxWMdXPZaTftRBPAU3N/NtpylTXsRu8X+sc/xQhs5Rm8QhWH/RLCzVdnusIG8q\nwywPecmXYrf4ZQyekGGH2LGvt3H9eApulQfQkwRDTvKSx2DDMfqwrNJno+3daN+vaNytmsRh2HzZ\nYN88z2MF+2qszKol+KQyVbbBco9PZGB20ZVYf8/3MUFZ0bF3n9wUbi3QMSmlbM0/u1KZ/JVKZXqr\nny+llCoTYMd3WlgsJfRUIR2lwWPTkFLfKa3iUgQuRdCMlDOBkxLO+QTZTI2YCUlSTgS+Q/IIpXXY\nG3ReJvBtiyQpJ2P9xvMTzpuPzZ90asJ5hSBJykE0P53Yz7F5mQpPkpTJNN8c8FJ0fuFJkrIRa+No\nhu0pyajrJCnLaX4u/EOB/2kvnHyQJOVO7BvlkITzDsZGY5di0aMkKVdjQ7nuAj7L0MUaR2LTPN+N\ntbVePdwBhiBJyjrsTtmMtc2+gc3R9CDW2LwO65vZHJ1X+MVeobm35IXYvLBzgT8G9o0dew6bvOoK\nGjf7FYpmmw5WY4ulnYf14m2PrcmxoUNxBaWV9pQNlFRGlWakTI3+fDn6cyQ2UrK2Jfx5bCHpwpMk\nZW+sI/0ibAwtWALlN8S572KFcOHrKknfPp/C5jD4ljj2RWx000zgo1gmx18Ma3SBSLpTZmIdVGrZ\nq0UMHt10CyUZApZ0p0zD6iPNsITA88gOF0l3yhjssYizFqur1E5v9iZDE6QKSZKUNxj49qnSx+DF\n1apMwYZ9FZ6kx+dJms9MP4qSTMuaJOU2bIKqpLbXE6Lzbms/pPAkSbkBK0BvwVZ42q3m+O7A30bH\nF9PeFPG5IalMeRsbgnUXVoG7ECtQ38Tef6qDd5ZE55Vi5blm+n1WYiMUz2Fg0M57GRioc3Z0vBQp\n/dD8C+FbWANSKRqRkvBuU4FLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FEEepNSm347Hltpb\nhg0wzHwmjjxIqV0B90uYlOnAfdG/MyW0FLUC7vEMpPLeSOP5KjtCaCkq/XYnBsa6rKHFTIx2CCml\nmfTbfgJMfhUy30el396M3R1TsIFCUzFxknopuPPm2e7e3t458fOLsjRwlXj67RXYwgVghWy9ebPr\n3kGVSqW/Uqk0c4e1ndbfaaoBXob1Ni7DRjFkPpl4XtLlfhltYB34RwaMJVd3Sm5wKQKXInApApci\ncCkClyJwKQKXInApApcicCkClyJwKQKXInApApcicCkClyJwKQKXInApApcicCkClyJwKQKXInAp\nApcicCkClyJwKQKXInApApcicCkClyJwKQKXInApApcicCkClyJwKQKXInApApciCC1lJLay7iIs\nDffSaP9+wMPAE9jUr2PlpztEaClvY0uD7o/NXHoE8PtYSu750b7bsYzUzAgtBQbW79gWy2Zfh81r\n+6to/73Y1PSZkQcpI7DHZw22lulT0VadxHM2sEuWAeUhh7APe3x2AO7BZis9Hfh74BKsTNmsPlj2\nFNwql2ATiseZjhXGtZQ2BXciA7NejMJSbx8HJkX7RgAXY+uRZUZoKVOB+7Ey5ddYwvZ9wBxgKTbH\n7QvYCruZEbpM6UUvPP/taAtC6Dsll7gUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsR\nuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgU\ngUsRuBSBSxG4FIFLEbgUgUsRuBRBaCn1UnBnAAuxhKj/BH43y6BC5/tUU3A3RrE8hKXgfhVLnbsH\nOBZbxfKIrIIKfaeATsFdjSVagqXTvZhlQKHvFLD/mMeAvbBcwaewNU0fAq6Mjn84y4DyIEWl4F4M\nnI1lr88GbsCSLgfRbSm4b8b29QDrxbldlYK7CFiOLRcMthLusiyDCv34TMXWWR8RbTdjcxt8FvgO\n8B7grejfmRFaSr0U3P8CDso4lt8Q+vHJJS5F4FIELkXgUgQuReBSBC5F4FIELkXgUgQuReBSBC5F\n4FIELkXgUgQuReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkVQailNjlEZQqmldCMt3QXNXMPvFIFL\nEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUQWgp9VJwb8HSbx8HVkV/puXwYYgvGKOj\nP7cGHsFScONcieUU1pLUyFRp4mfLa4ROgoKhKbhrY8d6gJPIMCcZwj8+MHQV3Kdjxw6N9q8IEFcu\n2AF7fA6P7bsWOLfO+Yuw27+dbdEw/w4dIb4K7tZYJvt7sw4i9ONTbxVcgCOxBV9fyjqo0AWtSsG9\nLzr2CeDHgeJyHIf3YYtUOzH2x2bdSEXogrZdvkzj6v7UVi7a01osuaEPe2HcWOf4KGAP7PWha1gB\nfLLB8ZYen9CVt3Z5HPhQ6CDyxgeBAxsc3wbYPZtQSk7RC9o4+wLTo78vA54IGEtwDsRm7Omr2Z4A\nDggYVzD2xmb6ehSYgxW6HwJOxuaOWx+d01X8GPgpuhgYER3rujftNTSe0Opg4JWMYskNm4CdGxzf\nJTonFUWvvL0A7Nfg+H5kPBlnHrgSm4Bzijg2NTp2VdqLFr2esiPwMPYI/RBr0wWr6c7B7pKDselc\nu4pxwDVYJ1q1jrI22jeulQsW/U6JMwKYFP39VVp4Oy4LY4FTGZjlOM4O0bHtMo0oB8wFftbg+L9h\nrXNdxaPArAbHj8Gq+13FemDXBsd3Q8963JCiV956gMkNjk+mhd+x6FKeAo5ucPzo6Jyu4tNYS/4J\n4tiJ0bHPZBpRDugBfoDVSRZjU8rfDiyN9nVds0Gck4B/xUZBLQbuiPY5Tn3uosXu0jKzAdiznQsU\n/Su5IxRdyv0MjJmr8hzwToBYckMfjWu0LVH0O6UjFH3QDliL/ciEc55Lc8Git7w107rWT8pBO2W4\nU45hcJJD1+MFbVYUXcqzwJbhvmjRC9o7sYJU/R7V/f3A8WkuWvSC9jjs63YBjeV0FZcDLwMrsTzD\nRiMQuoqtgY9jjUxvA3djCzpuEzKoPDEFuABYgnWdttQ7WPRvn1rGYN2lY7F2la5lNHAa8CDWev9D\n4KMhAwrN9QyMjvwcQ9tWWqLo9ZQ+4HkGBhLHv357Yvu6qp5yE1pEnK6rpziO4zhF5P8BQyU3ONKW\nGb0AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7feb58e17390>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "pj.report_seq_stats()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The sequences in the dataset are of different lengths. Sequences downloaded from Genbank (generated by Hou et al. 2007; Hou et al. 2011; Weiss et al. 2014; Copilas-Ciocianu & Petrusek 2015) are mostly full length COI barcoding fragments (Folmer et al. 1994). Sanger sequences produced by the current were also based on the full length COI barcode. Sequences obtained via eDNA metabarcoding approach are shorter (~313bp)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will generate three separate alignments.\n",
    " - 'ALL': All sequences from Genbank and current study (Sanger Folmer- and metabarcoding mini-barcode fragment)\n",
    " - 'FOLMER': Sequences from Genbank and current study Sanger Folmer fragments only\n",
    " - 'MINIBC': Sequences from Genbank and current study metabarcoding mini-barcode fragments only"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Configure and run sequence alignment 'ALL' with mafft."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --localpair --maxiterate 1000 287521484299993.59_MT-CO1.fasta\n"
     ]
    }
   ],
   "source": [
    "mafft_linsi = AlnConf(pj, \n",
    "                      method_name='ALL', \n",
    "                      CDSAlign=False, \n",
    "                      program_name='mafft', \n",
    "                      cmd='mafft', \n",
    "                      loci=['MT-CO1'], \n",
    "                      cline_args={'localpair': True, 'maxiterate': 1000})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Execute sequence alignment process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.align([mafft_linsi])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Display information about the alignment method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AlnConf named ALL with ID 287521484299993.59\n",
      "Loci: MT-CO1 \n",
      "Created on: Fri Jan 13 09:33:13 2017\n",
      "Commands:\n",
      "MT-CO1: mafft --localpair --maxiterate 1000 287521484299993.59_MT-CO1.fasta\n",
      "\n",
      "Environment:\n",
      "Platform: Linux-4.4.0-53-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.68\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.3\n",
      " User: chrishah-OptiPlex-7040\n",
      " Program and version: MAFFT v7.123b\n",
      " Program reference:Katoh\n",
      " Standley 2013 (Molecular Biology and Evolution 30:772-780) MAFFT multiple sequence alignment software version 7: improvements in performance and usability.\n",
      "execution time:\n",
      "17.9483218193\n",
      "\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "The dataset(s) MT-CO1 were aligned using the program MAFFT v7.123b [1].\n",
      "\n",
      "Reference:\n",
      "Katoh, Standley 2013 (Molecular Biology and Evolution 30:772-780) MAFFT multiple sequence alignment software version 7: improvements in performance and usability.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['ALL']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Produce the alignemnt 'FOLMER', i.e. remove all minibarcode sequences. These have the ids >= 'denovo5'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "denovo0 R_Taff_G.pulex_HCO2198\n",
      "denovo1 Little_Stour_G_fossarum_2_HCO2198\n",
      "denovo2 R_Taff_G_fossarum_1_HCO2198\n",
      "denovo3 R_Taff_G_fossarum_2_HCO2198\n",
      "denovo4 Little_Stour_G_fossarum_1_HCO2198\n",
      "denovo5 Gammarus_pulex|Gp_UK_Mb01\n",
      "denovo6 Gammarus_pulex|Gp_UK_Mb02\n",
      "denovo7 Gammarus_pulex|Gp_UK_Mb03\n",
      "denovo8 Gammarus_pulex|Gp_UK_Mb04\n",
      "denovo9 Gammarus_pulex|Gp_UK_Mb05\n",
      "denovo10 Gammarus_pulex|Gp_UK_Mb06\n",
      "denovo11 Gammarus_pulex|Gp_UK_Mb07\n",
      "denovo12 Gammarus_pulex|Gp_UK_Mb08\n",
      "denovo13 Gammarus_pulex|Gp_UK_Mb09\n",
      "denovo14 Gammarus_fossarum|Gf_UK_Mb01\n",
      "denovo15 Gammarus_fossarum|Gf_UK_Mb02\n",
      "denovo16 Gammarus_fossarum|Gf_UK_Mb03\n",
      "denovo17 Gammarus_fossarum|Gf_UK_Mb04\n",
      "denovo18 Gammarus_fossarum|Gf_UK_Mb05\n"
     ]
    }
   ],
   "source": [
    "for r in pj.records:\n",
    "    if 'denovo' in r.id:\n",
    "        print r.id,r.description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Empty sequence object\n",
    "seqobject = []\n",
    "\n",
    "#Add all records except the record ids >= 'denovo6' to new object\n",
    "for r in pj.records_by_locus['MT-CO1']:\n",
    "    if 'denovo' in r.id:\n",
    "        denovoID=int(r.id.replace('denovo','').replace('_f0',''))\n",
    "#        print denovoID\n",
    "        if denovoID >= 5:\n",
    "            pass\n",
    "        else:\n",
    "            seqobject.append(r)\n",
    "    else:\n",
    "        seqobject.append(r)\n",
    "\n",
    "#Replace original object with new \n",
    "pj.records_by_locus['MT-CO1'] = seqobject"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Configure and run alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --localpair --maxiterate 1000 647731484300507.53_MT-CO1.fasta\n"
     ]
    }
   ],
   "source": [
    "mafft_linsi = AlnConf(pj, \n",
    "                      method_name='FOLMER', \n",
    "                      CDSAlign=False, \n",
    "                      program_name='mafft', \n",
    "                      cmd='mafft', \n",
    "                      loci=['MT-CO1'], \n",
    "                      cline_args={'localpair': True, 'maxiterate': 1000})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.align([mafft_linsi])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Restore original alignemnt object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.extract_by_locus()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Produce the alignemnt 'MINIBC', i.e. remove all minibarcode sequences. These have ids <= 'denovo4'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Empty sequence object\n",
    "seqobject = []\n",
    "\n",
    "#Add all records except the record ids >= 'denovo6' to new object\n",
    "for r in pj.records_by_locus['MT-CO1']:\n",
    "    if 'denovo' in r.id:\n",
    "        denovoID=int(r.id.replace('denovo','').replace('_f0',''))\n",
    "#        print denovoID\n",
    "        if denovoID <= 4:\n",
    "            pass\n",
    "        else:\n",
    "            seqobject.append(r)\n",
    "    else:\n",
    "        seqobject.append(r)\n",
    "\n",
    "#Replace original object with new \n",
    "pj.records_by_locus['MT-CO1'] = seqobject"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Configure and run alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --localpair --maxiterate 1000 276191484300646.68_MT-CO1.fasta\n"
     ]
    }
   ],
   "source": [
    "mafft_linsi = AlnConf(pj, \n",
    "                      method_name='MINIBC', \n",
    "                      CDSAlign=False, \n",
    "                      program_name='mafft', \n",
    "                      cmd='mafft', \n",
    "                      loci=['MT-CO1'], \n",
    "                      cline_args={'localpair': True, 'maxiterate': 1000})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.align([mafft_linsi])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Restore original alignment object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.extract_by_locus()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write alignments to files for checking in e.g. `Aliview`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MT-CO1@MINIBC_aln.fasta', 'MT-CO1@FOLMER_aln.fasta', 'MT-CO1@ALL_aln.fasta']"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.write_alns(id=['record_id', 'source_organism'], format='fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Trim each of the three alignemnts to remove any positions that are not covered by at least 70, 80, and 90% of the records."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trimal -in 178771484300685.65_MT-CO1@MINIBC.fasta -gt 0.7\n",
      "trimal -in 178771484300685.65_MT-CO1@FOLMER.fasta -gt 0.7\n",
      "trimal -in 178771484300685.65_MT-CO1@ALL.fasta -gt 0.7\n"
     ]
    }
   ],
   "source": [
    "gt70trimal = TrimalConf(pj, \n",
    "                        method_name='gt70', \n",
    "                        program_name='trimal', \n",
    "                        cmd='trimal', \n",
    "                        alns=['MT-CO1@ALL', 'MT-CO1@FOLMER', 'MT-CO1@MINIBC'], \n",
    "                        trimal_commands={'gt': 0.7})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trimal -in 658571484300688.13_MT-CO1@MINIBC.fasta -gt 0.8\n",
      "trimal -in 658571484300688.13_MT-CO1@FOLMER.fasta -gt 0.8\n",
      "trimal -in 658571484300688.13_MT-CO1@ALL.fasta -gt 0.8\n"
     ]
    }
   ],
   "source": [
    "gt80trimal = TrimalConf(pj, \n",
    "                        method_name='gt80', \n",
    "                        program_name='trimal', \n",
    "                        cmd='trimal', \n",
    "                        alns=['MT-CO1@ALL', 'MT-CO1@FOLMER', 'MT-CO1@MINIBC'], \n",
    "                        trimal_commands={'gt': 0.8})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trimal -in 332591484300688.78_MT-CO1@MINIBC.fasta -gt 0.9\n",
      "trimal -in 332591484300688.78_MT-CO1@FOLMER.fasta -gt 0.9\n",
      "trimal -in 332591484300688.78_MT-CO1@ALL.fasta -gt 0.9\n"
     ]
    }
   ],
   "source": [
    "gt90trimal = TrimalConf(pj, \n",
    "                        method_name='gt90', \n",
    "                        program_name='trimal', \n",
    "                        cmd='trimal', \n",
    "                        alns=['MT-CO1@ALL', 'MT-CO1@FOLMER', 'MT-CO1@MINIBC'], \n",
    "                        trimal_commands={'gt': 0.9})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.trim([gt70trimal,gt80trimal,gt90trimal])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TrimalConf named gt90 with ID 332591484300688.78\n",
      "Alignments: MT-CO1@MINIBC,MT-CO1@FOLMER,MT-CO1@ALL \n",
      "Created on: Fri Jan 13 09:44:48 2017\n",
      "Commands:\n",
      "MT-CO1@FOLMER@gt90: trimal -in 332591484300688.78_MT-CO1@FOLMER.fasta -gt 0.9\n",
      "MT-CO1@ALL@gt90: trimal -in 332591484300688.78_MT-CO1@ALL.fasta -gt 0.9\n",
      "MT-CO1@MINIBC@gt90: trimal -in 332591484300688.78_MT-CO1@MINIBC.fasta -gt 0.9\n",
      "\n",
      "Environment:Platform: Linux-4.4.0-53-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.68\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.3\n",
      " User: chrishah-OptiPlex-7040\n",
      " Program and version: trimAl 1.2rev59\n",
      " Program reference: Salvador Capella-Gutierrez; Jose M. Silla-Martinez; Toni Gabaldon. trimAl: a tool for automated alignment trimming in large-scale phylogenetic analyses. Bioinformatics 2009 25: 1972-1973.\n",
      "execution time:\n",
      "0.764511108398\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "The alignment(s) MT-CO1@MINIBC;MT-CO1@FOLMER;MT-CO1@ALL were trimmed using the program trimAl 1.2rev59 [1].\n",
      "\n",
      "Reference:\n",
      "Salvador Capella-Gutierrez; Jose M. Silla-Martinez; Toni Gabaldon. trimAl: a tool for automated alignment trimming in large-scale phylogenetic analyses. Bioinformatics 2009 25: 1972-1973.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['gt90']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MT-CO1@MINIBC@gt80_trimmed_aln.fasta',\n",
       " 'MT-CO1@ALL@gt70_trimmed_aln.fasta',\n",
       " 'MT-CO1@MINIBC@gt90_trimmed_aln.fasta',\n",
       " 'MT-CO1@MINIBC@gt70_trimmed_aln.fasta',\n",
       " 'MT-CO1@ALL@gt80_trimmed_aln.fasta',\n",
       " 'MT-CO1@FOLMER@gt70_trimmed_aln.fasta',\n",
       " 'MT-CO1@ALL@gt90_trimmed_aln.fasta',\n",
       " 'MT-CO1@FOLMER@gt90_trimmed_aln.fasta',\n",
       " 'MT-CO1@FOLMER@gt80_trimmed_aln.fasta']"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.write_trimmed_alns(id=['record_id', 'source_organism'], format='fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Configure tree building."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@MINIBC@gt800 -p 566 -s 345621484300695.25_MT-CO1@MINIBC@gt80.fasta -T 3 -x 554 -N 100\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@ALL@gt700 -p 738 -s 345621484300695.25_MT-CO1@ALL@gt70.fasta -T 3 -x 744 -N 100\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@MINIBC@gt900 -p 168 -s 345621484300695.25_MT-CO1@MINIBC@gt90.fasta -T 3 -x 941 -N 100\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@MINIBC@gt700 -p 348 -s 345621484300695.25_MT-CO1@MINIBC@gt70.fasta -T 3 -x 826 -N 100\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@ALL@gt800 -p 894 -s 345621484300695.25_MT-CO1@ALL@gt80.fasta -T 3 -x 144 -N 100\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@FOLMER@gt700 -p 469 -s 345621484300695.25_MT-CO1@FOLMER@gt70.fasta -T 3 -x 352 -N 100\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@ALL@gt900 -p 418 -s 345621484300695.25_MT-CO1@ALL@gt90.fasta -T 3 -x 288 -N 100\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@FOLMER@gt900 -p 137 -s 345621484300695.25_MT-CO1@FOLMER@gt90.fasta -T 3 -x 917 -N 100\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@FOLMER@gt800 -p 830 -s 345621484300695.25_MT-CO1@FOLMER@gt80.fasta -T 3 -x 527 -N 100\n"
     ]
    }
   ],
   "source": [
    "raxml = RaxmlConf(pj, method_name='mafftLinsi-trimal', \n",
    "                 program_name='raxmlHPC-PTHREADS-SSE3',\n",
    "                 keepfiles=True,\n",
    "                 cmd='raxmlHPC-PTHREADS-SSE3',\n",
    "                 preset='fa',\n",
    "                 alns=['MT-CO1@FOLMER@gt80','MT-CO1@ALL@gt80','MT-CO1@MINIBC@gt80',\n",
    "                      'MT-CO1@FOLMER@gt70','MT-CO1@ALL@gt70','MT-CO1@MINIBC@gt70',\n",
    "                      'MT-CO1@FOLMER@gt90','MT-CO1@ALL@gt90','MT-CO1@MINIBC@gt90'],\n",
    "                 model='GAMMA',\n",
    "                 threads=3,\n",
    "                 cline_args={'-#': 100})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.tree([raxml])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RaxmlConf named mafftLinsi-trimal with ID 345621484300695.25\n",
      "Alignments: MT-CO1@MINIBC@gt80,MT-CO1@ALL@gt70,MT-CO1@MINIBC@gt90,MT-CO1@MINIBC@gt70,MT-CO1@ALL@gt80,MT-CO1@FOLMER@gt70,MT-CO1@ALL@gt90,MT-CO1@FOLMER@gt90,MT-CO1@FOLMER@gt80 \n",
      "Created on: Fri Jan 13 09:44:55 2017\n",
      "Commands:\n",
      "MT-CO1@MINIBC@gt80:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@MINIBC@gt800 -p 566 -s 345621484300695.25_MT-CO1@MINIBC@gt80.fasta -T 3 -x 554 -N 100\n",
      "MT-CO1@ALL@gt70:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@ALL@gt700 -p 738 -s 345621484300695.25_MT-CO1@ALL@gt70.fasta -T 3 -x 744 -N 100\n",
      "MT-CO1@MINIBC@gt90:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@MINIBC@gt900 -p 168 -s 345621484300695.25_MT-CO1@MINIBC@gt90.fasta -T 3 -x 941 -N 100\n",
      "MT-CO1@MINIBC@gt70:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@MINIBC@gt700 -p 348 -s 345621484300695.25_MT-CO1@MINIBC@gt70.fasta -T 3 -x 826 -N 100\n",
      "MT-CO1@ALL@gt80:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@ALL@gt800 -p 894 -s 345621484300695.25_MT-CO1@ALL@gt80.fasta -T 3 -x 144 -N 100\n",
      "MT-CO1@FOLMER@gt70:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@FOLMER@gt700 -p 469 -s 345621484300695.25_MT-CO1@FOLMER@gt70.fasta -T 3 -x 352 -N 100\n",
      "MT-CO1@ALL@gt90:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@ALL@gt900 -p 418 -s 345621484300695.25_MT-CO1@ALL@gt90.fasta -T 3 -x 288 -N 100\n",
      "MT-CO1@FOLMER@gt90:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@FOLMER@gt900 -p 137 -s 345621484300695.25_MT-CO1@FOLMER@gt90.fasta -T 3 -x 917 -N 100\n",
      "MT-CO1@FOLMER@gt80:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 345621484300695.25_MT-CO1@FOLMER@gt800 -p 830 -s 345621484300695.25_MT-CO1@FOLMER@gt80.fasta -T 3 -x 527 -N 100\n",
      "\n",
      "Environment:\n",
      "Platform: Linux-4.4.0-53-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.68\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.3\n",
      " User: chrishah-OptiPlex-7040\n",
      " Program and version: raxmlHPC-PTHREADS-SSE3RAxML version 8.0.12\n",
      " Program reference: A. Stamatakis: RAxML Version 8: A tool for Phylogenetic Analysis and Post-Analysis of Large Phylogenies. In Bioinformatics\n",
      " 2014.\n",
      "execution time:\n",
      "1588.31492305\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "Phylogenetic trees were reconstructed from the dataset(s) MT-CO1@MINIBC@gt80;MT-CO1@ALL@gt70;MT-CO1@MINIBC@gt90;MT-CO1@MINIBC@gt70;MT-CO1@ALL@gt80;MT-CO1@FOLMER@gt70;MT-CO1@ALL@gt90;MT-CO1@FOLMER@gt90;MT-CO1@FOLMER@gt80 using the program raxmlHPC-PTHREADS-SSE3RAxML version 8.0.12 [1].\n",
      "\n",
      "Reference:\n",
      "A. Stamatakis: RAxML Version 8: A tool for Phylogenetic Analysis and Post-Analysis of Large Phylogenies. In Bioinformatics, 2014.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['mafftLinsi-trimal']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pickle the project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'my_project.pickle'"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'my_project.pickle')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Annotate ML trees."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define outgroups."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "species\tMT-CO1\r",
      "\r\n",
      "Gammarus balcanicus\t1\r",
      "\r\n",
      "Gammarus fossarum\t98\r",
      "\r\n",
      "Gammarus glabratus\t1\r",
      "\r\n",
      "Gammarus pulex\t16\r",
      "\r\n",
      "Gammarus roeselii\t1\r",
      "\r\n",
      "Gammarus tigrinus\t1\r",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!cat species.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.add_qualifier_from_source('organism')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for species in ['balcanicus', 'glabratus', 'roeselii', 'tigrinus']:\n",
    "    pj.if_this_then_that(species, 'organism', 'outgroup', 'outgroup',\n",
    "                         mode='part')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify colors for branchsupport."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "supports = {'black': [100,95],\n",
    "            'dimgray': [95,75],\n",
    "            'silver': [75,50]}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add a new qualifier 'origin' that contains either 'Sanger' or 'metabarcoding' for the newly generated sequences. Also we add another qualifier 'origin_abr' that contains the symbols * or +, respectively.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for r in pj.records:\n",
    "    if 'denovo' in r.id:\n",
    "        print r.id\n",
    "        ## all the Sanger sequences contain the string 'HCO2198' in the sequence header\n",
    "        if 'HCO2198' in r.description:\n",
    "            print \"origin -> Sanger\"\n",
    "            for f in r.features:\n",
    "                f.qualifiers['origin'] = ['Sanger']\n",
    "                f.qualifiers['origin_abr'] = ['*']\n",
    "        ## if they are don't contain 'HCO2198' they originate from metabarcoding\n",
    "        else:\n",
    "            print \"origin -> metabarcoding\"\n",
    "            for f in r.features:\n",
    "                f.qualifiers['origin'] = ['metabarcoding']\n",
    "                f.qualifiers['origin_abr'] = ['+']\n",
    "    #if they are not 'denovo' then they were from a previous study\n",
    "    else:\n",
    "        print r.id\n",
    "        print \"origin -> previous study\"\n",
    "        for f in r.features:\n",
    "            f.qualifiers['origin'] = ['']\n",
    "            f.qualifiers['origin_abr'] = ['']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.propagate_metadata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Refine tree annotation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Format tree to add symbols * or + to leaf name to indicate 'Sanger' or 'metabarcoding' sequences, respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outroup', \n",
    "            ['source_organism', 'record_id','origin_abr'], \n",
    "            node_support_dict=supports, multifurc=50,\n",
    "            html='./mafftLinsi-trimal-raxml.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Color the tree labels according to origin of the sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "leaf_colors = {'Sanger':'red', 'metabarcoding': 'blue'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'mid', 'mid', \n",
    "            ['source_organism', 'record_id'], \n",
    "            node_support_dict=supports, multifurc=50, #None\n",
    "            leaf_node_color_meta = 'origin', leaf_label_colors = leaf_colors,\n",
    "            html='./mafftLinsi-trimal-raxml.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add haplotype abbreviations to records to be used as tree tip labels."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "MB_count={'Gp':1, 'Gf':1}\n",
    "FO_count={'Gp':1, 'Gf':1}\n",
    "for r in pj.records:\n",
    "    print r.id,r.features[0].qualifiers['organism'][0]\n",
    "\n",
    "    if r.features[0].qualifiers['organism'][0].split(\" \")[1] in ['pulex','fossarum']:\n",
    "        sp_prefix = r.features[0].qualifiers['organism'][0].split(\" \")[0][0]+r.features[0].qualifiers['organism'][0].split(\" \")[1][0]\n",
    "        if 'denovo' in r.id:\n",
    "            if 'metabarcoding' in r.features[0].qualifiers['origin']:\n",
    "                for f in r.features:\n",
    "                    f.qualifiers['label'] = [r.description.split(\"|\")[1]]\n",
    "                print r.features[0].qualifiers['label']\n",
    "                MB_count[sp_prefix]+=1\n",
    "            if 'Sanger' in r.features[0].qualifiers['origin']:\n",
    "                for f in r.features:\n",
    "                    f.qualifiers['label'] = [sp_prefix+'_UK_Fo'+\"%02d\" %FO_count[sp_prefix]]\n",
    "                print r.features[0].qualifiers['label']\n",
    "                FO_count[sp_prefix]+=1\n",
    "        elif r.id.startswith('KR'): #These are the records from Copilas-Ciocianu & Petrusek 2015\n",
    "            for f in r.features:\n",
    "                f.qualifiers['label'] = [sp_prefix+'_'+r.features[0].qualifiers['country'][0][:3]+'_'+r.features[0].qualifiers['isolate'][0]]\n",
    "            print r.features[0].qualifiers['label']\n",
    "        elif r.id.startswith('KF'): #These are the records from Weiss et al. 2014\n",
    "            for f in r.features:\n",
    "                f.qualifiers['label'] = r.features[0].qualifiers['specimen_voucher']\n",
    "            print r.features[0].qualifiers['label']\n",
    "        else: #These are the remaining records from Hou et al. 2007 and Hou et al. 2011\n",
    "            for f in r.features:\n",
    "                f.qualifiers['label'] = [sp_prefix+'_'+r.features[0].qualifiers['country'][0][:3]+'_'+r.features[0].qualifiers['specimen_voucher'][0]]\n",
    "            print r.features[0].qualifiers['label']\n",
    "    else:\n",
    "        for f in r.features:\n",
    "            f.qualifiers['label'] = r.features[0].qualifiers['organism']\n",
    "        print r.features[0].qualifiers['label']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.propagate_metadata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outgroup', \n",
    "            ['label'], \n",
    "            node_support_dict=supports, multifurc=None,\n",
    "            leaf_node_color_meta = 'origin', leaf_label_colors = leaf_colors,\n",
    "            html='./mafftLinsi-trimal-raxml.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write out alignments again, adding the new label names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MT-CO1@MINIBC_aln.fasta', 'MT-CO1@FOLMER_aln.fasta', 'MT-CO1@ALL_aln.fasta']"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.write_alns(id=['record_id', 'label'], format='fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MT-CO1@MINIBC@gt80_trimmed_aln.fasta',\n",
       " 'MT-CO1@ALL@gt70_trimmed_aln.fasta',\n",
       " 'MT-CO1@MINIBC@gt90_trimmed_aln.fasta',\n",
       " 'MT-CO1@MINIBC@gt70_trimmed_aln.fasta',\n",
       " 'MT-CO1@ALL@gt80_trimmed_aln.fasta',\n",
       " 'MT-CO1@FOLMER@gt70_trimmed_aln.fasta',\n",
       " 'MT-CO1@ALL@gt90_trimmed_aln.fasta',\n",
       " 'MT-CO1@FOLMER@gt90_trimmed_aln.fasta',\n",
       " 'MT-CO1@FOLMER@gt80_trimmed_aln.fasta']"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.write_trimmed_alns(id=['record_id', 'label'], format='fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Copy phylogenetic trees and underlying alignments in separate directory as supplement for the paper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mkdir supplementary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#copy and rename phylogenies\n",
    "for p in $(ls -1 *.png)\n",
    "do\n",
    "    base=$(echo $p | cut -d \"_\" -f 2-)\n",
    "    cp $p supplementary/RAxML_GTRGAMMA-$base\n",
    "done\n",
    "\n",
    "#copy alignments\n",
    "cp MT-CO1* supplementary/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!mkdir supplementary/RAxML/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!mv RAxML_* supplementary/RAxML/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pickle Reprophylo project for later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'my_project.pickle'"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'my_project.pickle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!cp my_project.pickle supplementary/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generate supplementary table linking haplotypes with accession numbers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence id\tspecies\tGenbank accession\treference\n",
      "Gammarus balcanicus\tGammarus balcanicus\tJF965914.1\tHou et al. (2011)\n",
      "Gammarus glabratus\tGammarus glabratus\tEF570307.1\tHou et al. 2009\n",
      "Gammarus roeselii\tGammarus roeselii\tJF965952.1\tHou et al. (2011)\n",
      "Gammarus tigrinus\tGammarus tigrinus\tFJ581688.1\tRadulovici et al. (2009)\n",
      "Gf_BF_N1\tGammarus fossarum\tKF521829.1\tWeiss et al. (2013)\n",
      "Gf_BF_N2\tGammarus fossarum\tKF521828.1\tWeiss et al. (2013)\n",
      "Gf_BF_N3\tGammarus fossarum\tKF521827.1\tWeiss et al. (2013)\n",
      "Gf_BF_S1\tGammarus fossarum\tKF521832.1\tWeiss et al. (2013)\n",
      "Gf_BF_S2\tGammarus fossarum\tKF521831.1\tWeiss et al. (2013)\n",
      "Gf_BF_S3\tGammarus fossarum\tKF521830.1\tWeiss et al. (2013)\n",
      "Gf_Cro5_01\tGammarus fossarum\tKF521806.1\tWeiss et al. (2013)\n",
      "Gf_Cro5_02\tGammarus fossarum\tKF521805.1\tWeiss et al. (2013)\n",
      "Gf_Cro8_01\tGammarus fossarum\tKF521804.1\tWeiss et al. (2013)\n",
      "Gf_Cro8_02\tGammarus fossarum\tKF521803.1\tWeiss et al. (2013)\n",
      "Gf_Cro8_03\tGammarus fossarum\tKF521802.1\tWeiss et al. (2013)\n",
      "Gf_Ger6_01\tGammarus fossarum\tKF521822.1\tWeiss et al. (2013)\n",
      "Gf_Ger6_02\tGammarus fossarum\tKF521821.1\tWeiss et al. (2013)\n",
      "Gf_Ger6_03\tGammarus fossarum\tKF521820.1\tWeiss et al. (2013)\n",
      "Gf_Rom_1BAF\tGammarus fossarum\tKR061776.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1BH15\tGammarus fossarum\tKR061778.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1BI11\tGammarus fossarum\tKR061781.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1BOC\tGammarus fossarum\tKR061784.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1BOG\tGammarus fossarum\tKR061785.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1C16\tGammarus fossarum\tKR061788.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1CA1V\tGammarus fossarum\tKR061789.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1CS101\tGammarus fossarum\tKR061792.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1CS102\tGammarus fossarum\tKR061795.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1CS12\tGammarus fossarum\tKR061797.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1CS15\tGammarus fossarum\tKR061800.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1CS17\tGammarus fossarum\tKR061802.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1DO2\tGammarus fossarum\tKR061805.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1GO25\tGammarus fossarum\tKR061842.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1GO4A\tGammarus fossarum\tKR061808.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1GO4B\tGammarus fossarum\tKR061810.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1LO2\tGammarus fossarum\tKR061812.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1MON\tGammarus fossarum\tKR061813.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1NE3\tGammarus fossarum\tKR061815.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1PC6\tGammarus fossarum\tKR061817.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1PL7\tGammarus fossarum\tKR061838.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1PLO\tGammarus fossarum\tKR061840.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1PM2\tGammarus fossarum\tKR061843.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1RE04\tGammarus fossarum\tKR061820.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1RE05\tGammarus fossarum\tKR061823.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1SIG3\tGammarus fossarum\tKR061825.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1TA25\tGammarus fossarum\tKR061827.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1TM104\tGammarus fossarum\tKR061830.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1VA5\tGammarus fossarum\tKR061833.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_1VA6\tGammarus fossarum\tKR061836.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2BH15\tGammarus fossarum\tKR061779.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2BI11\tGammarus fossarum\tKR061782.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2BOG\tGammarus fossarum\tKR061786.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2CA1V\tGammarus fossarum\tKR061790.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2CS101\tGammarus fossarum\tKR061793.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2CS102\tGammarus fossarum\tKR061796.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2CS12\tGammarus fossarum\tKR061798.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2CS15\tGammarus fossarum\tKR061801.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2CS17\tGammarus fossarum\tKR061803.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2DO2\tGammarus fossarum\tKR061806.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2GO4A\tGammarus fossarum\tKR061809.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2GO4B\tGammarus fossarum\tKR061811.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2MON\tGammarus fossarum\tKR061814.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2PC6\tGammarus fossarum\tKR061818.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2PL7\tGammarus fossarum\tKR061839.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2PLO\tGammarus fossarum\tKR061841.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2PM2\tGammarus fossarum\tKR061844.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2RE04\tGammarus fossarum\tKR061821.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2SIG3\tGammarus fossarum\tKR061826.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2TA25\tGammarus fossarum\tKR061828.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2TA4\tGammarus fossarum\tKR061846.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2TM104\tGammarus fossarum\tKR061831.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2VA5\tGammarus fossarum\tKR061834.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_2VA6\tGammarus fossarum\tKR061837.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3BAF\tGammarus fossarum\tKR061777.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3BH15\tGammarus fossarum\tKR061780.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3BI11\tGammarus fossarum\tKR061783.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3BOG\tGammarus fossarum\tKR061787.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3CA1V\tGammarus fossarum\tKR061791.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3CS101\tGammarus fossarum\tKR061794.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3CS12\tGammarus fossarum\tKR061799.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3CS17\tGammarus fossarum\tKR061804.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3DO2\tGammarus fossarum\tKR061807.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3NE3\tGammarus fossarum\tKR061816.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3PM2\tGammarus fossarum\tKR061845.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3RE04\tGammarus fossarum\tKR061822.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3RE05\tGammarus fossarum\tKR061824.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3TA25\tGammarus fossarum\tKR061829.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3TA4\tGammarus fossarum\tKR061847.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3TM104\tGammarus fossarum\tKR061832.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_3VA5\tGammarus fossarum\tKR061835.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Rom_CM17\tGammarus fossarum\tKR061819.1\tCopilas-Ciocianu & Petrusek (2015)\n",
      "Gf_Sl_O1\tGammarus fossarum\tKF521835.1\tWeiss et al. (2013)\n",
      "Gf_Sl_W1\tGammarus fossarum\tKF521833.1\tWeiss et al. (2013)\n",
      "Gf_Sl_W2\tGammarus fossarum\tKF521834.1\tWeiss et al. (2013)\n",
      "Gf_UK_Fo01\tGammarus fossarum\tKY464960\tcurrent study\n",
      "Gf_UK_Fo02\tGammarus fossarum\tKY464961\tcurrent study\n",
      "Gf_UK_Fo03\tGammarus fossarum\tKY464962\tcurrent study\n",
      "Gf_UK_Fo04\tGammarus fossarum\tKY464963\tcurrent study\n",
      "Gf_UK_Mb01\tGammarus fossarum\tKY464973\tcurrent study\n",
      "Gf_UK_Mb02\tGammarus fossarum\tKY464974\tcurrent study\n",
      "Gf_UK_Mb03\tGammarus fossarum\tKY464975\tcurrent study\n",
      "Gf_UK_Mb04\tGammarus fossarum\tKY464976\tcurrent study\n",
      "Gf_UK_Mb05\tGammarus fossarum\tKY464977\tcurrent study\n",
      "Gp_Fra_SLOCHN154\tGammarus pulex\tJF965941.1\tHou et al. (2011)\n",
      "Gp_Fra_SLOCHN155\tGammarus pulex\tJF965942.1\tHou et al. (2011)\n",
      "Gp_Ger_SLOCHN105\tGammarus pulex\tJF965940.1\tHou et al. (2011)\n",
      "Gp_Net_610\tGammarus pulex\tEF570334.1\tHou et al. (2007)\n",
      "Gp_Swe_SLOCHN022\tGammarus pulex\tJF965939.1\tHou et al. (2011)\n",
      "Gp_Swe_SLOCHN176\tGammarus pulex\tJF965943.1\tHou et al. (2011)\n",
      "Gp_UK_Fo01\tGammarus pulex\tKY464959\tcurrent study\n",
      "Gp_UK_Mb01\tGammarus pulex\tKY464964\tcurrent study\n",
      "Gp_UK_Mb02\tGammarus pulex\tKY464965\tcurrent study\n",
      "Gp_UK_Mb03\tGammarus pulex\tKY464966\tcurrent study\n",
      "Gp_UK_Mb04\tGammarus pulex\tKY464967\tcurrent study\n",
      "Gp_UK_Mb05\tGammarus pulex\tKY464968\tcurrent study\n",
      "Gp_UK_Mb06\tGammarus pulex\tKY464969\tcurrent study\n",
      "Gp_UK_Mb07\tGammarus pulex\tKY464970\tcurrent study\n",
      "Gp_UK_Mb08\tGammarus pulex\tKY464971\tcurrent study\n",
      "Gp_UK_Mb09\tGammarus pulex\tKY464972\tcurrent study\n",
      "\n"
     ]
    }
   ],
   "source": [
    "references = {}\n",
    "\n",
    "#previously published sequences used in the phylogeny\n",
    "refs = open('../1-download_references/phylogeny_refseqs.csv','r')\n",
    "refs.next()\n",
    "for r in refs:\n",
    "    cols=r.strip().split(\",\")\n",
    "    references[cols[1]] = cols[2]\n",
    "\n",
    "hts = {}   \n",
    "\n",
    "#read in Genbank accessions for sequences produced in this study from text file\n",
    "acc_fh = open('supplementary/this_study_ht_to_accessions.tsv')\n",
    "ht_to_new_accs = {}\n",
    "for ht in acc_fh:\n",
    "    ht_to_new_accs[ht.strip().split(\"\\t\")[0]] = ht.strip().split(\"\\t\")[1]\n",
    "\n",
    "#\n",
    "for r in pj.records:\n",
    "    hts[r.features[0].qualifiers['label'][0]] = \"%s\" %r.features[0].qualifiers['organism'][0]\n",
    "    \n",
    "    if \".\".join(r.id.split(\".\")[:-1]) in references:\n",
    "        hts[r.features[0].qualifiers['label'][0]]+= \"\\t%s\\t%s\" %(r.id, references[\".\".join(r.id.split(\".\")[:-1])])\n",
    "\n",
    "    else:\n",
    "        hts[r.features[0].qualifiers['label'][0]]+= \"\\t\"+ht_to_new_accs[r.features[0].qualifiers['label'][0]]+\"\\tcurrent study\"\n",
    "\n",
    "\n",
    "out='sequence id\\tspecies\\tGenbank accession\\treference\\n'\n",
    "for s in sorted(hts.keys()):\n",
    "    out+=\"%s\\t%s\\n\" %(s, hts[s])\n",
    "\n",
    "print out\n",
    "\n",
    "outfh=open('supplementary/ht_accessions.tsv','w')\n",
    "outfh.write(out)\n",
    "outfh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
